smcore/factory.cpp

```diff
@@ -22,19 +22,32 @@
 
 #include "factory.h"
 
+#include <map>
+
 namespace smos
 {
     namespace smcore
     {
         //******************************************************************************
         Options *Factory::getOptions(smos::smcore::SMString optionsfile)
         {
-            static Options *smOptions = nullptr;
-            if (nullptr == smOptions)
+            static std::map<smos::smcore::SMString, Options *> programOptions;
+            if (programOptions.find(optionsfile) == programOptions.end())
             {
-                smOptions = new Options(optionsfile);
+                programOptions.insert(std::make_pair(optionsfile, new Options(optionsfile)));
             }
-            return smOptions;
+            return programOptions[optionsfile];
         }
+        //******************************************************************************
+        Version *Factory::getVersion(void)
+        {
+            static Version *smVersion = nullptr;
+            if (nullptr == smVersion)
+            {
+                smVersion = new Version();
```

Comment:
We have a memory leak here. Object Version created on the heap will never be destroyed. Also, returning a raw pointer is not clear to me. Who is the owner of the created object? Is the caller of this method responsible for destroying the object? Knowing the implementation of this method, I can find out, but the fact of returning a raw pointer doesn't tell me that. If we really want to pass ownership of this object to the outside, we should use unique_ptr.
The correct version of this method, assuming we're still returning a raw pointer:
        Version *Factory::getVersion(void)
        {
            static Version smVersion;
            return &smVersion;
        }
All in all, I wonder about the purpose of this method.

Reply:
Uh, oh. Maybe I've understood something wrong. The basic idea is to have some kind of a singleton - the version object should be always and anywhere the same. And it should be alive until the program ends. Additional there should be only one class, where this kind of object should be created, in this case the factory class. But this is an acient concept and it's borrowed from C. Is there a similar concept in C++ available? And if yes, how does it looks like?